In [1]:
import pandas as pd
import regex as re

In [2]:
file_path = "episode_level_measures_parts_with_viewership_new_measures.csv"
# Read the CSV file
nlpdata = pd.read_csv(file_path, sep=',', on_bad_lines='skip', index_col=False, dtype='unicode')

# Display the first few rows of the dataframe to verify the content
nlpdata.head()

Show Season Episode  \
0  2 Broke Girls      1      01   
1  2 Broke Girls      1      02   
2  2 Broke Girls      1      03   
3  2 Broke Girls      1      04   
4  2 Broke Girls      1      06   

                                           File Name       sd_div_mean_1  \
0  2 Broke Girls - S01E01 - 2.Broke.Girls.Season....   5658.359420163232   
1  2 Broke Girls - S01E02 - 2 Broke Girls S01E02 ...   6280.502095337137   
2  2 Broke Girls - S01E03 - 2 Broke Girls S01E03 ...  3931.6040952853023   
3  2 Broke Girls - S01E04 - 2 Broke Girls S01E04 ...         5899.921656   
4  2 Broke Girls - S01E06 - 2 Broke Girls S01E06 ...   9570.497331194108   

       sd_div_mean_2       sd_div_mean_3            sd_sum_1  \
0        9159.676336   4626.990287281845   18.66514228373773   
1  3451.804293115786   7127.226693789984  19.733889186635707   
2  3977.461758366915   5506.511934027525   21.55941993523026   
3  7097.733100232231  3386.8579172793297   21.26736255478216   
4  5883.807212973031         6103.310464  20.229745482690152   

             sd_sum_2            sd_sum_3  ...            motion_1  \
0  20.545672701464397   21.69718205504922  ...  1.7342857142857144   
1  20.345822809580056   17.44018977836537  ...  1.4857142857142858   
2  20.759948099987184  19.837830499174945  ...   1.741428571428571   
3         20.24850059   19.43430838831536  ...   2.347142857142857   
4  21.176402608924068  21.713179209464133  ...  1.4257142857142855   

             motion_2            motion_3            space_1  \
0  2.4214285714285713              2.0725  5.192857142857143   
1   2.987142857142857  2.2185714285714284  6.852857142857142   
2  2.3614285714285717               5.745  6.055714285714286   
3  3.0485714285714285   2.847142857142857  7.688571428571429   
4                 1.9               2.015  6.475714285714285   

             space_2            space_3              time_1  \
0  7.394285714285714            5.41375   4.322857142857143   
1  7.992857142857143  6.631428571428572   4.492857142857143   
2  6.632857142857143            4.38625  4.8100000000000005   
3  6.484285714285714  6.192857142857142   4.957142857142857   
4               4.26            5.53875   4.502857142857143   

               time_2             time_3 Viewership (millions)  
0   3.005714285714286             2.5725                 19.37  
1   4.335714285714285  3.932857142857143                 11.75  
2   2.487142857142857            3.27375                 11.42  
3  3.7471428571428578  4.097142857142857                 10.71  
4                4.93            3.62375                 11.19  

[5 rows x 134 columns]

In [3]:
nlpdata.dtypes

Show                     object
Season                   object
Episode                  object
File Name                object
sd_div_mean_1            object
                          ...  
space_3                  object
time_1                   object
time_2                   object
time_3                   object
Viewership (millions)    object
Length: 134, dtype: object

In [4]:
# How many TV shows do we have
len(nlpdata.Show.unique())

235

In [5]:
nlpdata.loc[:,'sd_div_mean_1':'time_3'].shape

(26088, 129)

For each show, each episode we have sentiment analysis metrics + viewership (millions) of the debut release.
Each measure appears 3 times corresponding to different acts within an episode. We have 129 measures from the sentiment analysis output / 3 acts correspond to **43 emotions per act**.

***NLP Dictionary***:
- Sd_div_mean: Aims to assess how quickly a consumer becomes immersed in a story—both mentally and emotionally. Become absorbed in the narrative. By dividing the standard deviation of overall emotion scores by the mean, this measurement captures the shifts, variability, and intensity of emotions throughout the experience.
- Sd_sum: Sum of the standard deviation, amount of emotion variance for each act.
- Sd_scaled: Standard deviation scaled across all emotions for ease of comparison.
- Anger
- Surprise
- Disgust
- Sadness
- Neutral
- Fear
- Joy
- Positive
- Negative
- Engaged: High ssychological involvement or emotional investment, including greater use of personal pronouns, more emotional words, and greater cognitive processing contained in the act.
- Not Engaged: Low psychological involvement or emotional investment, including lesser use of personal pronouns, less emotional words, and less cognitive processing contained in the act.
- WC: Total word count contained in the act.
- Analytic: Analytical, formal, or logical discussion contained in the act.
- Clout: Social status, confidence, or leadership discussion contained in the act.
- Authenticity: Honest, non-filtered, non-regulated discussion contained in the act.
- Tone: The higher the tone, the more positive the tone in the act (below 50 is considered negative).
- WPS: Average words per sentence.
- Six letter: Percentage of words longer than six letters.
- Dic: Percentage of words that were captured as dictionary words.
- Cogprocess: An aggregate measurement that looks at the amount of words that reflect active information processing and mental activities, including causation contained in the act.
    - Insight: Considers realizations.
    - Cause: Examines causation between two elements.
    - Discrep: Considers what should, could, or would have happened, but never did, (exploring counterfactuals).
    - Tentative: Looks at whether something could or could not happen (e.g. maybe, perhaps).
    - Certain: Looks at absolute language (e.g. always, never).
    - Differ (Is this differentiate?): Considers differentiation between two elements such as (hasn’t , but, else).
- Perceptual: An aggregate measurement of terms that describe perception, such as look, heard, and feeling.
    - See: Amount of text around viewing or seeing.
    - Hear: Amount of text around hearing or listening.
    - Feel: Looks at references to touch or feeling.
- Drives: An aggregate measurement that looks at different motivations.
    - Affiliation: Looks at relations and affiliations such as ally, friend, or being social.
    - Achieve: Considers the ability to win, earn success, and be better.
    - Power: Examines power dynamics and structures including superiority and bullying.
    - Reward: Examines the types of rewards that are discussed including receiving something, prizes, and benefits.
    - Risk: Examines the different types of dangers and doubts.
- Relativity: Aggregate measure extends toward spatial relationships such as area, bend, and exit.
    - Motion: Examines the ability to move, including arrive, car, and go.
    - Space: Examines directions in space, including down, and in.
    - Time: Examines time durations, including end, until and season.


# Preprocessing

In [6]:
nlpdata.head()

Show Season Episode  \
0  2 Broke Girls      1      01   
1  2 Broke Girls      1      02   
2  2 Broke Girls      1      03   
3  2 Broke Girls      1      04   
4  2 Broke Girls      1      06   

                                           File Name       sd_div_mean_1  \
0  2 Broke Girls - S01E01 - 2.Broke.Girls.Season....   5658.359420163232   
1  2 Broke Girls - S01E02 - 2 Broke Girls S01E02 ...   6280.502095337137   
2  2 Broke Girls - S01E03 - 2 Broke Girls S01E03 ...  3931.6040952853023   
3  2 Broke Girls - S01E04 - 2 Broke Girls S01E04 ...         5899.921656   
4  2 Broke Girls - S01E06 - 2 Broke Girls S01E06 ...   9570.497331194108   

       sd_div_mean_2       sd_div_mean_3            sd_sum_1  \
0        9159.676336   4626.990287281845   18.66514228373773   
1  3451.804293115786   7127.226693789984  19.733889186635707   
2  3977.461758366915   5506.511934027525   21.55941993523026   
3  7097.733100232231  3386.8579172793297   21.26736255478216   
4  5883.807212973031         6103.310464  20.229745482690152   

             sd_sum_2            sd_sum_3  ...            motion_1  \
0  20.545672701464397   21.69718205504922  ...  1.7342857142857144   
1  20.345822809580056   17.44018977836537  ...  1.4857142857142858   
2  20.759948099987184  19.837830499174945  ...   1.741428571428571   
3         20.24850059   19.43430838831536  ...   2.347142857142857   
4  21.176402608924068  21.713179209464133  ...  1.4257142857142855   

             motion_2            motion_3            space_1  \
0  2.4214285714285713              2.0725  5.192857142857143   
1   2.987142857142857  2.2185714285714284  6.852857142857142   
2  2.3614285714285717               5.745  6.055714285714286   
3  3.0485714285714285   2.847142857142857  7.688571428571429   
4                 1.9               2.015  6.475714285714285   

             space_2            space_3              time_1  \
0  7.394285714285714            5.41375   4.322857142857143   
1  7.992857142857143  6.631428571428572   4.492857142857143   
2  6.632857142857143            4.38625  4.8100000000000005   
3  6.484285714285714  6.192857142857142   4.957142857142857   
4               4.26            5.53875   4.502857142857143   

               time_2             time_3 Viewership (millions)  
0   3.005714285714286             2.5725                 19.37  
1   4.335714285714285  3.932857142857143                 11.75  
2   2.487142857142857            3.27375                 11.42  
3  3.7471428571428578  4.097142857142857                 10.71  
4                4.93            3.62375                 11.19  

[5 rows x 134 columns]

In [8]:
nlpdata.dtypes

Show                     object
Season                   object
Episode                  object
File Name                object
sd_div_mean_1            object
                          ...  
space_3                  object
time_1                   object
time_2                   object
time_3                   object
Viewership (millions)    object
Length: 134, dtype: object

In [9]:
# Convert Season column to integer and Show column to string
nlpdata['Season'] = nlpdata['Season'].astype(int)
nlpdata['Show'] = nlpdata['Show'].astype(str)

# Identify columns to convert to float (excluding 'Show', 'Season', and 'Episode')
columns_to_convert = nlpdata.columns.difference(['Show', 'Season', 'Episode'])
nlpdata[columns_to_convert] = nlpdata[columns_to_convert].apply(pd.to_numeric, errors='coerce')

# Drop unnecessary columns
nlpdata.drop(columns = ['File Name'], inplace=True)

# Clean 'Episode' column
nlpdata['Episode'] = (
    nlpdata['Episode']
    .str.strip()  # Remove leading/trailing whitespace
    .str.split('-')  # Take the first part before '-'
    .str[0]
    .str.extract(r'(\d+)')[0]  # Keep only digits
    .astype(int)
)

nlpdata.head()

Show  Season  Episode  sd_div_mean_1  sd_div_mean_2  \
0  2 Broke Girls       1        1    5658.359420    9159.676336   
1  2 Broke Girls       1        2    6280.502095    3451.804293   
2  2 Broke Girls       1        3    3931.604095    3977.461758   
3  2 Broke Girls       1        4    5899.921656    7097.733100   
4  2 Broke Girls       1        6    9570.497331    5883.807213   

   sd_div_mean_3   sd_sum_1   sd_sum_2   sd_sum_3  sd_scaled_1  ...  motion_1  \
0    4626.990287  18.665142  20.545673  21.697182   598.633647  ...  1.734286   
1    7127.226694  19.733889  20.345823  17.440190   629.842904  ...  1.485714   
2    5506.511934  21.559420  20.759948  19.837830   694.243568  ...  1.741429   
3    3386.857917  21.267363  20.248501  19.434308   684.387650  ...  2.347143   
4    6103.310464  20.229745  21.176403  21.713179   648.841555  ...  1.425714   

   motion_2  motion_3   space_1   space_2   space_3    time_1    time_2  \
0  2.421429  2.072500  5.192857  7.394286  5.413750  4.322857  3.005714   
1  2.987143  2.218571  6.852857  7.992857  6.631429  4.492857  4.335714   
2  2.361429  5.745000  6.055714  6.632857  4.386250  4.810000  2.487143   
3  3.048571  2.847143  7.688571  6.484286  6.192857  4.957143  3.747143   
4  1.900000  2.015000  6.475714  4.260000  5.538750  4.502857  4.930000   

     time_3  Viewership (millions)  
0  2.572500                  19.37  
1  3.932857                  11.75  
2  3.273750                  11.42  
3  4.097143                  10.71  
4  3.623750                  11.19  

[5 rows x 133 columns]

In [10]:
# Show data
imdb_file_path = "show_info_pivoted_with_imdb (1).csv"
# Read the CSV file
imdb = pd.read_csv(imdb_file_path, sep=',', on_bad_lines='skip', index_col=False, dtype='unicode')

#View data2
imdb.head()

Show                                Executive producers  \
0          2 Broke Girls  Michael Patrick King | Whitney Cummings | Mich...   
1                     24  Joel Surnow | Robert Cochran | Brian Grazer | ...   
2                30 Rock  Lorne Michaels | Tina Fey | Marci Klein | Davi...   
3  3rd Rock from the Sun  John Lithgow | Bonnie Turner | Terry Turner | ...   
4               9/1/2001  Ryan Murphy | Brad Falchuk | Tim Minear | Juan...   

                                               Genre    Network  \
0                                             Sitcom        CBS   
1  Serial drama | Crime thriller | Espionage | Ac...        Fox   
2                            Sitcom | Satire | Farce        NBC   
3                           Sitcom | Science fiction        NBC   
4                                   Procedural drama  Fox | ABC   

                                     No.of episodes No.of seasons  \
0                                               138             6   
1  192 + 24: Redemption + 12 (24: Live Another Day)             9   
2                                               139             7   
3                                               139             6   
4                                                96             6   

                                            Starring  Year  
0  Kat Dennings | Beth Behrs | Garrett Morris | J...  2011  
1              Kiefer Sutherland | (and | others | )  2001  
2  Tina Fey | Tracy Morgan | Jane Krakowski | Jac...  2006  
3  John Lithgow | Kristen Johnston | French Stewa...  1996  
4  Angela Bassett | Peter Krause | Oliver Stark |...  2018

In [11]:
print(imdb.shape)
imdb.dtypes

(304, 8)


Show                   object
Executive producers    object
Genre                  object
Network                object
No.of episodes         object
No.of seasons          object
Starring               object
Year                   object
dtype: object

In [12]:
# Filter rows where 'Episode' contains non-numeric characters
invalid_episodes = imdb[imdb['No.of episodes'].apply(lambda x: not str(x).isdigit())]
invalid_episodes[['Show', 'No.of episodes']]

Show  \
1                          24   
35            Brooklyn Bridge   
43       Caroline in the City   
45                     Cheers   
60                     Cursed   
72          Diagnosis: Murder   
92                Family Ties   
95                   Fired Up   
111                     Grand   
119              Hearts Afire   
130  In the Heat of the Night   
132           Inside Schwartz   
138            Just Shoot Me!   
140               Key & Peele   
154                Love & War   
165                   Matlock   
169             Modern Family   
173         Murder, She Wrote   
187           Out of Practice   
203              Room for Two   
210            Schitt's Creek   
229          Stark Raving Mad   
233            Suddenly Susan   
242            The Cosby Show   
246       The Eric Andre Show   
260           The Naked Truth   
269            The Single Guy   
281                  Top Gear   
288              Union Square   
292      Walker, Texas Ranger   

                                        No.of episodes  
1     192 + 24: Redemption + 12 (24: Live Another Day)  
35                   35 (includes 2 two-part episodes)  
43                                      97 (1 unaired)  
45   275 (including three double-length episodes an...  
60                                      17 (2 unaired)  
72                           178 + 5 TV movies + Pilot  
92                                      176 + one film  
95                                      28 (5 unaired)  
111                                     26 (1 unaired)  
119                                     54 (1 unaired)  
130                                  147 + 4 TV movies  
132                                     13 (4 unaired)  
138            145 (+ 3 episodes aired in syndication)  
140                                 53 (and 1 special)  
154                                     68 (5 unaired)  
165                                  193 + pilot movie  
169                                   250 (+1 special)  
173                                264 (+ 4 TV movies)  
187                                     22 (8 unaired)  
203                                     26 (2 unaired)  
210                                    80 (+1 special)  
229                                     22 (4 unaired)  
233                                     93 (1 unaired)  
242                        201 (plus outtakes special)  
246                               60 (plus 3 specials)  
260                                     55 (7 unaired)  
269                              43 (+1 Unaired Pilot)  
281                        240 (including 13 specials)  
288                                     14 (1 unaired)  
292                           203 (not incl. TV movie)

In [13]:
# Convert relevant columns to appropriate types
imdb['Year'] = imdb['Year'].astype(int)

# Define the function to truncate after the first number
def truncate_after_first_number(text):
    match = re.search(r'\d+', str(text))  # Search for the first occurrence of one or more digits
    return match.group() if match else "0"  # Return the matched number as a string or '0' if no number is found

# Step 1: Handle non-numeric entries in 'No.of seasons'
imdb['No.of seasons'] = imdb['No.of seasons'].apply(truncate_after_first_number).astype(int)
# Step 2: Handle non-numeric entries in 'No.of episodes' and convert to integer
imdb['No.of episodes'] = imdb['No.of episodes'].apply(truncate_after_first_number).astype(int)

imdb.head()

Show                                Executive producers  \
0          2 Broke Girls  Michael Patrick King | Whitney Cummings | Mich...   
1                     24  Joel Surnow | Robert Cochran | Brian Grazer | ...   
2                30 Rock  Lorne Michaels | Tina Fey | Marci Klein | Davi...   
3  3rd Rock from the Sun  John Lithgow | Bonnie Turner | Terry Turner | ...   
4               9/1/2001  Ryan Murphy | Brad Falchuk | Tim Minear | Juan...   

                                               Genre    Network  \
0                                             Sitcom        CBS   
1  Serial drama | Crime thriller | Espionage | Ac...        Fox   
2                            Sitcom | Satire | Farce        NBC   
3                           Sitcom | Science fiction        NBC   
4                                   Procedural drama  Fox | ABC   

   No.of episodes  No.of seasons  \
0             138              6   
1             192              9   
2             139              7   
3             139              6   
4              96              6   

                                            Starring  Year  
0  Kat Dennings | Beth Behrs | Garrett Morris | J...  2011  
1              Kiefer Sutherland | (and | others | )  2001  
2  Tina Fey | Tracy Morgan | Jane Krakowski | Jac...  2006  
3  John Lithgow | Kristen Johnston | French Stewa...  1996  
4  Angela Bassett | Peter Krause | Oliver Stark |...  2018

In [17]:
# Check again for invalid episodes
invalid_episodes = imdb[imdb['No.of episodes'].apply(lambda x: not str(x).isdigit())]
invalid_episodes[['Show', 'No.of episodes']]

Empty DataFrame
Columns: [Show, No.of episodes]
Index: []

In [15]:
#Drop irrelevant columns
imdb.drop(columns = ["Executive producers", "Starring"], errors='ignore', axis=1, inplace=True)

# Extract the first genre before '|'
imdb['Genre'] = imdb['Genre'].str.split('|').str[0]
# Extract the first network before '|'
imdb['Network'] = imdb['Network'].str.split('|').str[0]

In [16]:
imdb.head()

Show             Genre Network  No.of episodes  \
0          2 Broke Girls            Sitcom     CBS             138   
1                     24     Serial drama      Fox             192   
2                30 Rock           Sitcom      NBC             139   
3  3rd Rock from the Sun           Sitcom      NBC             139   
4               9/1/2001  Procedural drama    Fox               96   

   No.of seasons  Year  
0              6  2011  
1              9  2001  
2              7  2006  
3              6  1996  
4              6  2018

In [18]:
# One-hot encoding of Network column
imdb['Network'] = imdb['Network'].str.strip()  # Clean the 'Network' column
imdb_encoded = pd.concat([imdb.drop('Network', axis=1), 
    pd.get_dummies(imdb['Network'], prefix='Network')], axis=1)

In [19]:
len(imdb.Show.unique())

304

In [20]:
imdb_encoded.head()

Show             Genre  No.of episodes  No.of seasons  \
0          2 Broke Girls            Sitcom             138              6   
1                     24     Serial drama              192              9   
2                30 Rock           Sitcom              139              7   
3  3rd Rock from the Sun           Sitcom              139              6   
4               9/1/2001  Procedural drama              96              6   

   Year  Network_ABC  Network_AMC  Network_Adult Swim  Network_BBC Two  \
0  2011        False        False               False            False   
1  2001        False        False               False            False   
2  2006        False        False               False            False   
3  1996        False        False               False            False   
4  2018        False        False               False            False   

   Network_CBC Television  ...  Network_HBO  Network_HBO (US)  Network_IFC  \
0                   False  ...        False             False        False   
1                   False  ...        False             False        False   
2                   False  ...        False             False        False   
3                   False  ...        False             False        False   
4                   False  ...        False             False        False   

   Network_NBC  Network_Paramount Network  Network_Showtime  Network_Starz  \
0        False                      False             False          False   
1        False                      False             False          False   
2         True                      False             False          False   
3         True                      False             False          False   
4        False                      False             False          False   

   Network_The CW  Network_The WB  Network_USA Network  
0           False           False                False  
1           False           False                False  
2           False           False                False  
3           False           False                False  
4           False           False                False  

[5 rows x 27 columns]

In [21]:
# Merge data and imdb_enconded dataframes on 'Show'
nlpdata_imdb = pd.merge(nlpdata, imdb_encoded, on='Show', how='left')
nlpdata_imdb.head()

Show  Season  Episode  sd_div_mean_1  sd_div_mean_2  \
0  2 Broke Girls       1        1    5658.359420    9159.676336   
1  2 Broke Girls       1        2    6280.502095    3451.804293   
2  2 Broke Girls       1        3    3931.604095    3977.461758   
3  2 Broke Girls       1        4    5899.921656    7097.733100   
4  2 Broke Girls       1        6    9570.497331    5883.807213   

   sd_div_mean_3   sd_sum_1   sd_sum_2   sd_sum_3  sd_scaled_1  ...  \
0    4626.990287  18.665142  20.545673  21.697182   598.633647  ...   
1    7127.226694  19.733889  20.345823  17.440190   629.842904  ...   
2    5506.511934  21.559420  20.759948  19.837830   694.243568  ...   
3    3386.857917  21.267363  20.248501  19.434308   684.387650  ...   
4    6103.310464  20.229745  21.176403  21.713179   648.841555  ...   

   Network_HBO  Network_HBO (US)  Network_IFC  Network_NBC  \
0        False             False        False        False   
1        False             False        False        False   
2        False             False        False        False   
3        False             False        False        False   
4        False             False        False        False   

   Network_Paramount Network  Network_Showtime  Network_Starz  Network_The CW  \
0                      False             False          False           False   
1                      False             False          False           False   
2                      False             False          False           False   
3                      False             False          False           False   
4                      False             False          False           False   

   Network_The WB  Network_USA Network  
0           False                False  
1           False                False  
2           False                False  
3           False                False  
4           False                False  

[5 rows x 159 columns]

In [36]:
len(nlpdata_imdb.Show.unique())

235

In [176]:
# Display a statistical summary of the 'Viewership (millions)' column
nlpdata_imdb['Viewership (millions)'].describe().T

count    25861.000000
mean        10.667185
std          7.817323
min          0.049000
25%          4.830000
50%          9.080000
75%         14.900000
max         58.530000
Name: Viewership (millions), dtype: float64

In [177]:
#Based on this, it makes sense to drop rows that have missing values
nlpdata_imdb.dropna(inplace=True)

In [41]:
file_path3 = "Air_Date1.csv"
airdate_data = pd.read_csv(file_path3, sep=',', index_col=False, dtype='unicode')

# Splitting Episode ranges into separate rows e.g. episode 21-22 into 2 separate rows
airdate_data_split = pd.concat([pd.DataFrame({
    'Show': row['Show'],
    'Season': row['Season'],
    'Episode': episode,
    'Air Date': row['Air Date']
}, index=[0]) for _, row in airdate_data.iterrows() for episode in row['Episode'].split('-') if episode.isdigit()]).reset_index(drop=True)

# Convert 'Season' and 'Episode' columns to integers
airdate_data_split[['Season','Episode']] = airdate_data_split[['Season','Episode']].astype(int)

airdate_data_split.head()

Show  Season  Episode   Air Date
0  2 Broke Girls       1        1  19-Sep-11
1  2 Broke Girls       1        2  26-Sep-11
2  2 Broke Girls       1        3   3-Oct-11
3  2 Broke Girls       1        4  10-Oct-11
4  2 Broke Girls       1        5  17-Oct-11

In [55]:
len(airdate_data_split.Show.unique())
airdate_data_split.dtypes

Show        object
Season       int32
Episode      int32
Air Date    object
dtype: object

In [51]:
# Merge merged_data and data3_split based on 'Show', 'Season', and 'Episode'
merged_df = pd.merge(nlpdata_imdb, airdate_data_split, on=['Show', 'Season', 'Episode'], how='outer')
merged_df.head()

Show  Season  Episode  sd_div_mean_1  sd_div_mean_2  \
0  2 Broke Girls       1        1    5658.359420    9159.676336   
1  2 Broke Girls       1        2    6280.502095    3451.804293   
2  2 Broke Girls       1        3    3931.604095    3977.461758   
3  2 Broke Girls       1        4    5899.921656    7097.733100   
4  2 Broke Girls       1        5            NaN            NaN   

   sd_div_mean_3   sd_sum_1   sd_sum_2   sd_sum_3  sd_scaled_1  ...  \
0    4626.990287  18.665142  20.545673  21.697182   598.633647  ...   
1    7127.226694  19.733889  20.345823  17.440190   629.842904  ...   
2    5506.511934  21.559420  20.759948  19.837830   694.243568  ...   
3    3386.857917  21.267363  20.248501  19.434308   684.387650  ...   
4            NaN        NaN        NaN        NaN          NaN  ...   

   Network_HBO (US)  Network_IFC  Network_NBC  Network_Paramount Network  \
0             False        False        False                      False   
1             False        False        False                      False   
2             False        False        False                      False   
3             False        False        False                      False   
4               NaN          NaN          NaN                        NaN   

   Network_Showtime  Network_Starz  Network_The CW  Network_The WB  \
0             False          False           False           False   
1             False          False           False           False   
2             False          False           False           False   
3             False          False           False           False   
4               NaN            NaN             NaN             NaN   

   Network_USA Network   Air Date  
0                False  19-Sep-11  
1                False  26-Sep-11  
2                False   3-Oct-11  
3                False  10-Oct-11  
4                  NaN  17-Oct-11  

[5 rows x 160 columns]

In [111]:
# Function to check for invalid dates
def check_invalid_dates(df_date_column, date_format):
    invalid_indices = []
    for index, date_str in enumerate(df_date_column):
        try:
            pd.to_datetime(date_str, format=date_format)
        except ValueError:
            invalid_indices.append(index)
    return invalid_indices

# Check for invalid dates in 'Air Date' column
invalid_indices = check_invalid_dates(merged_df['Air Date'], '%d-%b-%y')
print(len(invalid_indices), 'rows with invalid \'Air Date\' values')
merged_df['Air Date'].iloc[invalid_indices]

29 rows with invalid 'Air Date' values


889                  Aired in syndication
890                  Aired in syndication
891                  Aired in syndication
1015                 "Career Day, Part 2"
4458                              Unaired
10103                             Unaired
10104                             Unaired
10105                             Unaired
10106                             Unaired
10107                             Unaired
11702                             Unaired
12759                             Unaired
16548                             Unaired
16549                             Unaired
16550                             Unaired
16551                             Unaired
16552                             Unaired
17318                             Unaired
21749                             Unaired
21750                             Unaired
24299                             Unaired
25489    Holly Hester & Apryl Huntzinger 
26703                             Unaired
26704                             

In [112]:
# Drop rows with invalid dates
merged_df.drop(index=invalid_indices, inplace=True)

# Another way to check ivalid and drop
#merged_df['Air Date'] = pd.to_datetime(merged_df['Air Date'], format='%d-%b-%y', errors='coerce')
#merged_df = merged_df.dropna(subset=['Air Date'])

In [113]:
# Convert 'Air Date' column to datetime with correct format
merged_df['Air Date'] = pd.to_datetime(merged_df['Air Date'], format='%d-%b-%y')
merged_df.head()

Show  Season  Episode  sd_div_mean_1  sd_div_mean_2  \
0  2 Broke Girls       1        1    5658.359420    9159.676336   
1  2 Broke Girls       1        2    6280.502095    3451.804293   
2  2 Broke Girls       1        3    3931.604095    3977.461758   
3  2 Broke Girls       1        4    5899.921656    7097.733100   
4  2 Broke Girls       1        5            NaN            NaN   

   sd_div_mean_3   sd_sum_1   sd_sum_2   sd_sum_3  sd_scaled_1  ...  \
0    4626.990287  18.665142  20.545673  21.697182   598.633647  ...   
1    7127.226694  19.733889  20.345823  17.440190   629.842904  ...   
2    5506.511934  21.559420  20.759948  19.837830   694.243568  ...   
3    3386.857917  21.267363  20.248501  19.434308   684.387650  ...   
4            NaN        NaN        NaN        NaN          NaN  ...   

   Network_HBO (US)  Network_IFC  Network_NBC  Network_Paramount Network  \
0             False        False        False                      False   
1             False        False        False                      False   
2             False        False        False                      False   
3             False        False        False                      False   
4               NaN          NaN          NaN                        NaN   

   Network_Showtime  Network_Starz  Network_The CW  Network_The WB  \
0             False          False           False           False   
1             False          False           False           False   
2             False          False           False           False   
3             False          False           False           False   
4               NaN            NaN             NaN             NaN   

   Network_USA Network   Air Date  
0                False 2011-09-19  
1                False 2011-09-26  
2                False 2011-10-03  
3                False 2011-10-10  
4                  NaN 2011-10-17  

[5 rows x 160 columns]

In [114]:
print(merged_df.shape)
# Drop rows with missing values
merged_df_cleaned = merged_df.dropna()
print(merged_df_cleaned.shape)

(30606, 160)
(705, 160)


In [118]:
# Many missing values in sentiment analysis and Network fields
merged_df.isna().sum().head(20)

Show                 0
Season               0
Episode              0
sd_div_mean_1    29901
sd_div_mean_2    29901
sd_div_mean_3    29901
sd_sum_1         29901
sd_sum_2         29901
sd_sum_3         29901
sd_scaled_1      29901
sd_scaled_2      29901
sd_scaled_3      29901
anger_1          29901
anger_2          29901
anger_3          29901
surprise_1       29901
surprise_2       29901
surprise_3       29901
disgust_1        29901
disgust_2        29901
dtype: int64

In [119]:
merged_df.isna().sum().tail(20)

Network_BBC Two                  29901
Network_CBC Television           29901
Network_CBS                      29901
Network_Cartoon Network          29901
Network_Comedy Central           29901
Network_Disney Channel           29901
Network_FX                       29901
Network_First-run syndication    29901
Network_Fox                      29901
Network_HBO                      29901
Network_HBO (US)                 29901
Network_IFC                      29901
Network_NBC                      29901
Network_Paramount Network        29901
Network_Showtime                 29901
Network_Starz                    29901
Network_The CW                   29901
Network_The WB                   29901
Network_USA Network              29901
Air Date                             0
dtype: int64

In [121]:
# Filter Shows based on minimum number of episodes threshold
min_episodes_threshold = 5  # Adjust as needed

# Calculate number of episodes per show
episode_counts = data.groupby('Show').size()
# Filter out shows with fewer episodes than the threshold
shows_to_keep = episode_counts[episode_counts >= min_episodes_threshold].index

# Filter the original DataFrame based on shows to keep
filtered_data = data[data['Show'].isin(shows_to_keep)]

In [122]:
shows_to_keep

Index(['2 Broke Girls', '24', '30 Rock', '3rd Rock from the Sun',
       '9-1-1: Lone Star', '9/1/2001', 'A Different World',
       'A Million Little Things', 'Abbott Elementary', 'Adventure Time',
       ...
       'Two and a Half Men', 'Ugly Betty', 'Veep', 'Walker, Texas Ranger',
       'Weeds', 'What We Do in the Shadows', 'Without a Trace',
       'Yellowjackets', 'Yellowstone', 'Young Sheldon'],
      dtype='object', name='Show', length=229)

In [129]:
merged_df['Genre'].unique()

array(['Sitcom', nan, 'Serial drama ', 'Sitcom ', 'Procedural drama'],
      dtype=object)

In [58]:
# Step 1: Load the new file
cancelled = pd.read_csv('Cancelled coding - NLP Survival Analysis.csv')
# Step 2: Merge the DataFrames on the 'Show' column
combined_df = pd.merge(merged_df, cancelled, on='Show', how='inner')
# Step 3: Optionally, inspect the merged DataFrame
combined_df.head()

Show  Season  Episode  sd_div_mean_1  sd_div_mean_2  \
0  2 Broke Girls       1        1    5658.359420    9159.676336   
1  2 Broke Girls       1        2    6280.502095    3451.804293   
2  2 Broke Girls       1        3    3931.604095    3977.461758   
3  2 Broke Girls       1        4    5899.921656    7097.733100   
4  2 Broke Girls       1        6    9570.497331    5883.807213   

   sd_div_mean_3   sd_sum_1   sd_sum_2   sd_sum_3  sd_scaled_1  ...  \
0    4626.990287  18.665142  20.545673  21.697182   598.633647  ...   
1    7127.226694  19.733889  20.345823  17.440190   629.842904  ...   
2    5506.511934  21.559420  20.759948  19.837830   694.243568  ...   
3    3386.857917  21.267363  20.248501  19.434308   684.387650  ...   
4    6103.310464  20.229745  21.176403  21.713179   648.841555  ...   

   Network_IFC  Network_NBC  Network_Paramount Network  Network_Showtime  \
0        False        False                      False             False   
1        False        False                      False             False   
2        False        False                      False             False   
3        False        False                      False             False   
4        False        False                      False             False   

   Network_Starz  Network_The CW  Network_The WB  Network_USA Network  \
0          False           False           False                False   
1          False           False           False                False   
2          False           False           False                False   
3          False           False           False                False   
4          False           False           False                False   

    Air Date  Cancelled  
0  19-Sep-11          1  
1  26-Sep-11          1  
2   3-Oct-11          1  
3  10-Oct-11          1  
4  24-Oct-11          1  

[5 rows x 161 columns]

In [132]:
combined_df['Air Date']

0      2011-09-19
1      2011-09-26
2      2011-10-03
3      2011-10-10
4      2011-10-24
          ...    
700    2023-04-17
701    2023-04-24
702    2023-05-01
703    2023-05-08
704    2023-05-15
Name: Air Date, Length: 705, dtype: object

# TV Series Regression

In [ ]:
#TV Series 1st
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Initialize list to store OLS results
Show_OLS_results_list = []

# TV series regressions
for show in merged_df_cleaned['Show'].unique():
        # Filter data for the current show
    show_data = merged_df_cleaned[merged_df_cleaned['Show'] == show]
    
    # Shift the viewership column to lag it by one episode
    show_data.loc[:, 'Viewership (millions)'] = show_data['Viewership (millions)'].shift(-1)
    
    # Drop the last row since it will have a NaN value in the lagged viewership column
    show_data = show_data.dropna()
    
    # Continue with the rest of your analysis
    if len(show_data) < 5:  # Adjust this number based on your minimum required samples
        print(f"Skipping show '{show}' due to insufficient samples (n={len(show_data)})")
        continue  # Skip to the next show
    
    # Drop unnecessary columns for regression
    X = show_data.drop(["Show", "Season", "Episode", "Viewership (millions)", "Network_USA Network"], axis=1, errors='ignore')
    y = show_data['Viewership (millions)']
    
    # Ensure X contains only numeric columns
    X = X.select_dtypes(include=[np.number])
    
    # Verify if X and y are not empty
    if X.empty or y.empty:
        print(f"Empty data in features or target variable for show '{show}'.")
        continue
   
    # Track feature names before scaling
    original_feature_names = X.columns
    
    # Scale the features
    scaler = MinMaxScaler()
    X_scaled = scaler.fit_transform(X)
    
    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)
    
    # Feature selection using RandomForestRegressor
    model_rf = RandomForestRegressor(n_estimators=100, random_state=42)
    
    try:
        model_rf.fit(X_train, y_train)
    except ValueError as e:
        print(f"Error fitting RandomForestRegressor for show '{show}': {e}")
        continue
    
    # Get feature importances and select top 10 features
    feature_importances = model_rf.feature_importances_
    top_indices = np.argsort(feature_importances)[-10:]  # Indices of top 10 features
    selected_features = X.columns[top_indices]
    
    # Select only the important features
    X_selected = X_scaled[:, top_indices]
    
    # Now add the VIF calculation and removal step here
    # Create a DataFrame for the selected features
    X_selected_df = pd.DataFrame(X_selected, columns=selected_features)
    
    # Function to calculate VIF and remove features with VIF > 10
    def calculate_vif(X, threshold=10):
        while True:
            vif_data = pd.DataFrame()
            vif_data["Feature"] = X.columns
            vif_data["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
            
            # Check if all VIF values are below the threshold
            max_vif = vif_data["VIF"].max()
            if max_vif < threshold:
                break
            
            # Find the feature with the highest VIF and drop it
            feature_to_drop = vif_data.loc[vif_data["VIF"].idxmax(), "Feature"]
            print(f"Dropping '{feature_to_drop}' with VIF: {max_vif}")
            X = X.drop(columns=[feature_to_drop])
        
        return X, vif_data

    # Apply the VIF function to filter out high VIF features
    X_selected_vif_filtered, vif_data = calculate_vif(X_selected_df)
    
    # If fewer than 10 features remain, re-add features with the lowest VIF or highest importance
    if X_selected_vif_filtered.shape[1] < 10:
        remaining_features_count = X_selected_vif_filtered.shape[1]
        missing_count = 10 - remaining_features_count

        # Find the remaining features from the original top features that were removed
        removed_features = [feature for feature in selected_features if feature not in X_selected_vif_filtered.columns]

        # Sort removed features based on their importance
        sorted_removed_features = sorted(removed_features, key=lambda x: feature_importances[X.columns.get_loc(x)], reverse=True)

        # Re-add features to reach 10
        features_to_add = sorted_removed_features[:missing_count]
        X_selected_vif_filtered = pd.concat([X_selected_vif_filtered, X_selected_df[features_to_add]], axis=1)

    # Update X_selected with filtered features
    X_selected = X_selected_vif_filtered.values
    selected_features = X_selected_vif_filtered.columns
    
    # Calculate correlation matrix with selected features
    correlation_matrix = pd.DataFrame(X_selected, columns=selected_features).corr()
  
    # Perform linear regression with selected features
    X_selected_train, X_selected_test, y_train, y_test = train_test_split(X_selected, y, test_size=0.2, random_state=42)
    X_selected_train = sm.add_constant(X_selected_train)  # Add a constant for the intercept

    model = sm.OLS(y_train, X_selected_train).fit()

    # Extract coefficients, p-values, and other metrics (this section remains the same)
    coefficients = model.params
    p_values = model.pvalues
    
    # Get confidence intervals
    conf = model.conf_int()  # This returns the confidence intervals for all coefficients
    
    # Compute F-statistic manually
    dof_model = X_selected_train.shape[1] - 1  # degrees of freedom of the model
    dof_resid = len(y_train) - X_selected_train.shape[1]  # degrees of freedom of residuals
    ssr = np.sum(model.resid ** 2)  # sum of squared residuals
    mse_resid = ssr / dof_resid  # mean squared error of residuals
    msr = np.sum((model.fittedvalues - np.mean(y_train)) ** 2) / dof_model  # mean squared error of regression
    f_statistic = msr / mse_resid  # F-statistic
    
    # Compute robust standard errors - heteroskedasticity
    robust_results = model.get_robustcov_results(cov_type='HC3')  # HC3 is one of the robust options
    
    # Predictions
    y_pred = model.predict(sm.add_constant(X_selected_test))
    
    # Calculate regression metrics (same as before)
    intercept = model.params[0]
    r_squared = model.rsquared
    adj_r_squared = model.rsquared_adj
    mae = mean_absolute_error(y_test, y_pred)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    
    # Cross-Validation
    ols_model = LinearRegression()
    cv_scores = cross_val_score(ols_model, X_selected_vif_filtered, y, cv=5, scoring='r2')
    cv_mean_score = np.mean(cv_scores)
    
    # Create a heatmap for the correlation matrix
    plt.figure(figsize=(10, 8))
    sns.heatmap(correlation_matrix, annot=True, fmt=".2f", cmap="coolwarm",
                xticklabels=correlation_matrix.columns,
                yticklabels=correlation_matrix.columns)
    plt.title(f"Correlation Matrix Heatmap for Show {show}")
    plt.show()
    
    # Display the correlation matrix
    print(f"Correlation matrix for show {show} with selected features:\n")
    print(correlation_matrix)
    print('\n' + '='*80 + '\n')
    
    # Print OLS regression summary
    print(f"OLS Regression results for show '{show}':\n")
    print(model.summary())
    
    # Create a DataFrame for regression results
    results_df = pd.DataFrame({
        'Feature': ['Intercept'] + list(selected_features),  # Include 'Intercept' and selected features
        'Coefficient': coefficients,
        'P-value': p_values,
        'CI Lower Bound': conf[0],
        'CI Upper Bound': conf[1]
    })
    # Append additional regression metrics
    metrics_df = pd.DataFrame({
        'Metric': ['Intercept', 'R-squared', 'Adjusted R-squared', 'Mean Absolute Error', 'Root Mean Square Error', 'F-statistic', 'F-statistic P-value'],
        'Value': [coefficients[0], model.rsquared, model.rsquared_adj, mae, rmse, f_statistic, model.f_pvalue],
})
       
    # Perform cross-validation for regression
    kf = KFold(n_splits=5, shuffle=True, random_state=42)
    cv_scores = cross_val_score(LinearRegression(), X_selected, y, cv=kf, scoring='r2')
    
    # Print cross-validation scores
    print(f"Cross-validation scores for show {show}:\n")
    print(cv_scores)
    print(f"Mean CV score: {cv_scores.mean()}")
    print('\n' + '='*80 + '\n')
    from statsmodels.stats.outliers_influence import variance_inflation_factor

    # Calculate VIF for the selected features
    vif_data = pd.DataFrame()
    vif_data["Feature"] = selected_features
    vif_data["VIF"] = [variance_inflation_factor(X_selected, i) for i in range(X_selected.shape[1])]

    # Print regression results
    print(f"Regression results for show {show} with selected features:\n")
    print(results_df.to_string(index=False))
    print('\n' + '='*80 + '\n')

    # Print metrics results
    print(f"Metrics results for show {show} with selected features:\n")
    print(metrics_df.to_string(index=False))
    print('\n' + '='*80 + '\n')
    
    # Store OLS results
    Show_OLS_results_list.append({
        'Show': show,
        'OLS_R_squared': r_squared,
        'OLS_Adj_R_squared': adj_r_squared,
        'OLS_MAE': mae,
        'OLS_RMSE': rmse,
        'CV_Mean_Score': cv_scores.mean()
    })

# Convert results to DataFrame and save to Excel
results_df = pd.DataFrame(Show_OLS_results_list)
results_df.to_excel('Show_OLS_results_list.xlsx', index=False)